### PR TITLE
ci: improve the release docs workflow

### LINKED ISSUE #191: 🛠 CI: release_docs/update_and_deploy_docs is failing
### Bug Report Checklist

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have pulled the latest `main` branch of the repository.
- [X] I have [searched for related issues](https://github.com/JoshuaKGoldberg/ts-api-utils/issues?q=is%3Aissue) and found none that matched my issue.

### Overview

https://github.com/JoshuaKGoldberg/ts-api-utils/actions/runs/4581824676/jobs/8091630871

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by RebeccaStevens**
## PR Checklist

- [x] Addresses an existing open issue: fixes #191
- [x] That issue was marked as [`status: accepting prs`](https://github.com/JoshuaKGoldberg/ts-api-utils/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/JoshuaKGoldberg/ts-api-utils/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

The same issue as https://github.com/RebeccaStevens/uom-types/issues/55

## Before Merge

`gh-pages` branch need to be created. Create with: 

```sh
git switch --orphan gh-pages
git commit --allow-empty -m "initial docs" --no-verify
git push
```

### CODE REVIEW

**Comment by RebeccaStevens** [comment]
Hmm, not sure if there's a nice way to ignore those spelling issues other than just adding the detected words to the cspell dictionary.

File: package.json [hunk 1]
```diff
@@ -28,7 +28,9 @@
 	],
 	"scripts": {
 		"build": "tsup src/index.ts --clean --format cjs,esm --outDir lib --dts && cp lib/index.d.ts lib/index.d.cts",
-		"docs": "typedoc",
+		"docs": "pnpm run docs:build && pnpm run docs:build-post",
+		"docs:build": "typedoc",
+		"docs:build-post": "find -lname \"$PWD/docs/generated/*\" | while read link; do target=$(readlink $link); relative_target=$(realpath $target --relative-to=$link/.. --no-symlinks --canonicalize-missing); ln --force --symbolic --no-target-directory $relative_target $link; done",
```
Comment (JoshuaKGoldberg) [inline package.json:33]: Uhh, this is a lot of shell scripting and magics. What's going on here?

Comment (RebeccaStevens) [reply package.json:33]: All it does is convert the absolute symlinks to relative symlinks. I open an issue here to see if we can get this done nicer: https://github.com/citkane/typedoc-plugin-versions/issues/29

**Comment by RebeccaStevens** [comment]
Pushed a new update the rework the workflows a bit. They aren't fully tested as testing workflows is hard, but hopefully they work as intended.

File: .github/workflows/docs-update.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          ref: gh-pages
+
+      - name: Save old docs
+        run: tar -cvf /tmp/old-docs.tar ./docs/generated
+
+      - name: Checkout tag
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ github.ref_name }}
+          clean: false #  doesn't seem to work?? Hense use of tar.
```
Comment (JoshuaKGoldberg) [inline .github/workflows/docs-update.yml:37]: [Question] What happened when you tried?

File: .github/workflows/docs-update.yml [hunk 3]
```diff
@@ -0,0 +1,70 @@
+name: Update Docs
+
+on:
+  push:
+    tags:
+      # https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string -- github doesn't like this regex??
```
Comment (JoshuaKGoldberg) [inline .github/workflows/docs-update.yml:6]: [Question] What about not liking it?

**Review by JoshuaKGoldberg — APPROVED** [review]
Cool! This is getting a bit complicated and long term I think it'd be nice to extract the logic out into its own reusable workflows. But for now it feels reasonable to just go with it. Thanks!

File: package.json [hunk 4]
```diff
@@ -28,7 +28,9 @@
 	],
 	"scripts": {
 		"build": "tsup src/index.ts --clean --format cjs,esm --outDir lib --dts && cp lib/index.d.ts lib/index.d.cts",
-		"docs": "typedoc",
+		"docs": "pnpm run docs:build && pnpm run docs:build-post",
+		"docs:build": "typedoc",
+		"docs:build-post": "sh scripts/build-post.sh",
```
Comment (JoshuaKGoldberg) [inline package.json:33]: `lint:spelling` (cspell) was failing on a couple of words in the script. I also don't like putting a lot of logic in `package.json`s - it gets unweildy. So I moved it to a separate file. WDYT @RebeccaStevens?

**Comment by JoshuaKGoldberg** [comment]
👋 ping @RebeccaStevens, is this something you're interested in and have time for?

**Comment by RebeccaStevens** [comment]
What's the status of this? What needs doing?

**Comment by JoshuaKGoldberg** [comment]
I'd left a few questions for you in comment threads. I'm up for merging as-is with the suggested touchups, but didn't want to steamroll past you. 🙂

File: .github/workflows/docs-update.yml (hunk 3 above)
Comment (RebeccaStevens) [reply .github/workflows/docs-update.yml:6]: Not sure, it just doesn't work.
Splitting it into two separate regexps works though.

File: .github/workflows/docs-update.yml (hunk 2 above)
Comment (RebeccaStevens) [reply .github/workflows/docs-update.yml:37]: The same thing as when I don't

File: package.json (hunk 4 above)
Comment (RebeccaStevens) [reply package.json]: `docs:build-post` no longer needed that to update in `typedoc-plugin-versions`

**Comment by RebeccaStevens** [comment]
I've rebased this off #341
This PR should be pretty much ready now.

**Review by JoshuaKGoldberg — APPROVED** [review]
Cool, thanks! Let's give this a shot. 🚀

**Pull request merged**